// webmake/build.js
(function (modules) {
	var getBuild = function (build) {
		return function (ignore, module) {
			module.exports = build.exports;
		};
	};
	var getModule = function (scope, tree, path) {
		var name, dir, exports = {}, module = { exports: exports }, require, build;
		path = path.split('/');
		name = path.pop();
		while ((dir = path.shift())) {
			if (dir === '..') {
				scope = tree.pop();
			} else if (dir !== '.') {
				tree.push(scope);
				scope = scope[dir];
			}
		}
		require = getRequire(scope, tree);
		build = scope[name];
		scope[name] = getBuild(module);
		build.call(exports, exports, module, require);
		return module.exports;
	};
	var require = function (scope, tree, path) {
		var name;
		if (path.charAt(0) !== '.') {
			name = path.split('/', 1)[0];
			path = path.slice(name.length + 1);
			scope = modules[name];
			tree = [];
			if (!path) {
				path = scope[':mainpath:'];
			}
		}
		return getModule(scope, tree, path);
	};
	var getRequire = function (scope, tree) {
		return function (path) {
			return require(scope, [].concat(tree),
				(path.slice(-3) === '.js') ? path.slice(0, -3) : path);
		};
	};
	return getRequire(modules['.'], []);
})
({
	".": {
		"lib": {
			"program": function (exports, module, require) {
				exports.x = require("./x");
				exports.y = require("./y.js");
				exports.outer = require("../outer");

				exports.external = {
					main:  require('test'),
					other: require('test/lib/other.js'),
					noMain: require('no-main/lib/some-module')
				};
			},
			"x": function (exports, module, require) {
				module.exports = {
					name: "x",
					getZ: function () {
						return require("./z");
					}
				};
			},
			"y": function (exports, module, require) {
				exports.name = "y";
				exports.z = require("./z");
			},
			"z": function (exports, module, require) {
				exports.name = "z";
				exports.y = require("./y");
			}
		},
		"outer": function (exports, module, require) {
			'use strict';

			module.exports = {
				name: "outer"
			};
		}
	},
	"test": {
		":mainpath:": "lib/chosen-one",
		"lib": {
			"chosen-one": function (exports, module, require) {
				'use strict';

				exports.name = "external-main";
				exports.module = require('./module');
			},
			"other": function (exports, module, require) {
				'use strict';

				exports.name = "external-other";
			},
			"module": function (exports, module, require) {
				'use strict';

				exports.name = 'module';
			}
		}
	},
	"no-main": {
		"lib": {
			"some-module": function (exports, module, require) {
				exports.name = 'no-main';			}
		}
	}
})
("./lib/program");
